pkg/engine/shield/add.go

```diff
@@ -4,20 +4,46 @@ import (
 	"github.com/simimpact/srsim/pkg/engine/event"
 	"github.com/simimpact/srsim/pkg/engine/info"
 	"github.com/simimpact/srsim/pkg/key"
+	"github.com/simimpact/srsim/pkg/model"
 )
 
 func (mgr *Manager) AddShield(id key.Shield, shield info.Shield) {
 	// TODO: LOGIC FOR ADDING A SHIELD
 	// 1. Check if the target already has this shield, if so remove old (mgr.RemoveShield)
+	if mgr.HasShield(shield.Target, id) {
+		mgr.RemoveShield(id, shield.Target)
+	}
+
 	// 2. Get the stats for the source and target
-	// 3. Compute shield HP/create ShieldInstance given the add paramsm
+	// 3. Compute shield HP/create ShieldInstance given the add params
+	shieldHP := 0.0
+
+	for k, v := range shield.BaseShield {
+		switch k {
+		case model.ShieldFormula_SHIELD_BY_SHIELDER_ATK:
+			shieldHP += v * mgr.attr.Stats(shield.Source).ATK()
+		case model.ShieldFormula_SHIELD_BY_SHIELDER_DEF:
+			shieldHP += v * mgr.attr.Stats(shield.Source).DEF()
+		case model.ShieldFormula_SHIELD_BY_SHIELDER_MAX_HP:
+			shieldHP += v * mgr.attr.Stats(shield.Source).HP()
+		case model.ShieldFormula_SHIELD_BY_TARGET_MAX_HP:
+			shieldHP += v * mgr.attr.Stats(shield.Target).HP()
+		case model.ShieldFormula_SHIELD_BY_SHIELDER_TOTAL_SHIELD:
+			shieldHP += shield.ShieldValue
+		}
+	}
+
+	newInstance := &Instance{name: id, HP: shieldHP}
```

Comment:
missing the use of the `ShieldTaken` and `ShieldAbsorb` properties: https://github.com/simimpact/srsim/blob/12711171278f35df30a278955b67435af4708556/pkg/engine/prop/prop.go#L60-L64

These properties buff the shield HP being applied. These work similar to Heal: https://github.com/simimpact/srsim/blob/12711171278f35df30a278955b67435af4708556/pkg/engine/combat/heal.go#L52-L55